patches/net/minecraft/world/level/block/ComposterBlock.java.patch

```diff
@@ -0,0 +1,19 @@
+--- a/net/minecraft/world/level/block/ComposterBlock.java
++++ b/net/minecraft/world/level/block/ComposterBlock.java
+@@ -219,6 +_,16 @@
+       if (p_51978_.getValue(LEVEL) == 7) {
+          p_51979_.scheduleTick(p_51980_, p_51978_.getBlock(), 20);
+       }
++      // NeoForge: Invalidate composter capabilities when added or removed
++      if (!p_51981_.is(this)) p_51979_.invalidateCapabilities(p_51980_);
++   }
++
++   @Override
++   public void onRemove(BlockState p_60515_, Level p_60516_, BlockPos p_60517_, BlockState p_60518_, boolean p_60519_) {
++      super.onRemove(p_60515_, p_60516_, p_60517_, p_60518_, p_60519_);
++      // NeoForge: Invalidate composter capabilities when added or removed
++      // TODO: why do we have two places where we call invalidateCapabilities?
```

Comment:
Not sure why this TODO is here, but from what I can tell it has to do with the fact there is no general capability invalidation on a spot when a block change occurs which means each capability provider effectively has to invalidate on creation and removal.

Does this potentially cause issues for caps that might be attached to vanilla blocks via the event? At the very least we probably want to document this behavior somewhere as I think as it stands currently modders would be required to also listen to block break/change events to mark that capabilities have been invalidated in that position.

Also the comment should use the standard `Neo:` not `NeoForge:`

Reply:
> Does this potentially cause issues for caps that might be attached to vanilla blocks via the event? At the very least we probably want to document this behavior somewhere as I think as it stands currently modders would be required to also listen to block break/change events to mark that capabilities have been invalidated in that position.

We have this javadoc already:
![image](https://github.com/neoforged/NeoForge/assets/13494793/053de800-e405-4aa0-8deb-3e248f670dfd)

Is this not clear enough? Personally I think you should never register a capability to a block that you don't own because of the invalidation problems. We don't have events that cover arbitrary block state changes. (Unless you mixin to add proper invalidation)

Reply:
Yeah I suppose the answer would have to basically be that you need to coremod to invalidate if you invalidate more than just on break/place.

And I think I may not have actually checked what the docs on the provider were before writing that part of the comment (oops)